### DISCUSSION TITLE
Cold snapshot GNSS / snapshot positioning without initial information

### DISCUSSION THREAD

**Discussion by JonasBchrt** [JonasBchrt/snapshot-gnss-algorithms#2] [Q&A] [answered]
I received a question regarding the cold snapshot GNSS implementation in this repository, which I would like to address here.

**Comment by JonasBchrt** [comment, marked as answer]
The cold snapshot technique combines **coarse-time Doppler navigation (CTDN)** with code-phase-based **coarse-time navigation (CTN)**. Applying CTDN first removes the requirement of CTN for a coarse initial receiver location and time. [My implementation](https://github.com/JonasBchrt/snapshot-gnss-algorithms/blob/422d679237b6aa0b891943a0b1bab59fc85e5dfe/coarse_time_navigation.py#L793) in _coarse_time_navigation.py_ is called `cold_snapshot(...)` and inspired by

> Fernández-Hernández, I., Borre, K. Snapshot positioning without initial information. _GPS Solut_ **20**, 605–616 (2016). https://doi.org/10.1007/s10291-016-0530-4.

Here are a few notes that might help understanding challenges when applying my code in practice:

- There was indeed a bug in the cold snapshot GNSS implementation in this repository. I think I fixed it in commit 422d679237b6aa0b891943a0b1bab59fc85e5dfe. Thanks for the heads-up.
- I wrote the `cold_snapshot(...)` routine before we even built our first SnapperGPS receiver and it has actually never been used for SnapperGPS. Cold snapshot GNSS is useful when you have no idea about initial values, but this is not relevant to SnapperGPS since we always have a rough user-provided initial position and a sufficiently accurate on-board real-time clock. Therefore, I did not spend time on robustifying or accelerating the cold snapshot implementation or optimising it for the use of multiple GNSS, unlike for some other algorithms. In consequence, I would not be surprised if my cold snapshot implementation is less reliable than some of the other algorithms in this repository when applied to real-world data.
- I created [a GitHub repo with a working cold snapshot GNSS example](https://github.com/JonasBchrt/cold-snapshot-demo). It uses a (heavily) down-sampled version of the data that Fernández-Hernández and Borre label with _ID 2_ in their paper.
- If you want to apply `cold_snapshot(...)` to a different dataset, then I would like to make you aware of the following aspect of CTDN: If the intermediate frequency (IF) of the raw signal is almost zero (or almost a multiple of the sampling frequency), then Doppler-based positioning is challenging. A standard acquisition routine cannot distinguish between a shift of -200 Hz and +200 Hz. So, you need to figure out the sign in the subsequent optimisation, too. My `cold_snapshot(...)` does not account for this; so, if you feed data like this into it, then it will almost certainly fail. I had to deal with the same issue in [ _frequency_bias_estimation.py_](https://github.com/JonasBchrt/snapshot-gnss-algorithms/blob/main/frequency_bias_estimation.py), but for a different purpose. I am robustly estimating the frequency offset of the RF front-end there, which is just one of the unknowns in the CTDN problem. The second half of Section 4.1 in our SnapperGPS paper covers this

> Jonas Beuchert and Alex Rogers. 2021. SnapperGPS: Algorithms for Energy-Efficient Low-Cost Location Estimation Using GNSS Signal Snapshots. In SenSys ’21: ACM Conference on Embedded Networked Sensor Systems, November, 2021, Coimbra, Portugal. ACM, New York, NY, USA, 13 pages. https://doi.org/10.1145/3485730.3485931.

- If there are any further questions regarding this topic, feel free to ask.

**Reply by willhakkrd** [reply]
Thanks a lot for this information.
However, I have 2 additional questions:
How is it that the standard acquisition cannot distinguish between -200 and +200 Hz?
Why the need to (heavily) downsample the data? I tried to change it by setting the decimation factor to 1, but apart from it taking a long time to process, it gave a wrong solution?

Thanks in advance!

**Reply by JonasBchrt** [reply]
There is no need to down-sample the data. I just did it because it makes the data more similar to the data that we get from our SnapperGPS receivers. However, [my demo script](https://github.com/JonasBchrt/cold-snapshot-demo/blob/master/demo_cold_snapshot.py) did not make it particularly easy to use different sampling rates. You had to change the value of the variables `decimation_factor` and `intermediate_frequency`, potentially the sign of the `acquired_doppler` array, and potentially the acquisition parameter `snr_threshold`. I simplified it in commit [fd4683e](https://github.com/JonasBchrt/cold-snapshot-demo/commit/fd4683edf5634c624c206abea919a06595f72aa1) such that you only have to change the `decimation_factor` now. In theory, it might still be necessary to change the `snr_threshold`, too, but the value that I have selected now seems to work for decimation factors 1 and 7. (In general, it is expected that the signal-to-noise ratio changes when data is down-sampled, which can require a different threshold for good satellites. However, the routine does not actually use the SNR, but a normalised value, which allows for similar thresholds across sampling frequencies, but might still require some tuning. SnapperGPS does not use such a hard threshold at all and dynamically selects good satellites, but this is not implemented in the cold snapshot algorithm.)

Still, some decimation factors might not work because the intermediate frequency would be too close to zero.

**Reply by JonasBchrt** [reply]
Can you detail your question "_How is it that the standard acquisition cannot distinguish between -200 and +200 Hz?_" A simple answer would be the following: Because of the fundamental property of the cosine that `cos(x) = cos(-x)`, a signal with a negative frequency is the same as the signal with a positive frequency if the absolute values of the frequencies are the same. (If you do not consider the phase of the signal, which we do not do in the acquisition routine.) However, I do not think that this is what you are asking for.

**Reply by ZiglioUK** [reply]
This is not true, maybe you're confusing non coherent acquisition with signal squaring. Of course acquisition gives you a doppler estimate across the whole doppler search space, including its sign. If you demodulate from -200Hz and you find a peak that's the doppler. You won't find it for the same signal if you demodulate from 200Hz. That's in standard acquisition we're talking about.

**Reply by JonasBchrt** [reply]
Thanks for your comment. A 2nd pair of eyes might minimise the change that I confuse more than I help in this forum.

Maybe I did not make clear that I was not talking about Doppler shifts of -200 Hz and +200 Hz, but about signals whose frequencies deviate -200 Hz and +200 Hz from an intermediate frequency (centre frequency) of 0 Hz. I do not think you can distinguish the signals in the latter case in a GNSS snapshot of a few milliseconds.

But if you think that I am wrong here, I would be happy about an explanation or a link.

Below is a code snippet that generates a GNSS-like "input signal" with 0 Hz centre frequency and +2 kHz shift. Then, it compares this "input signal" with signal replicas with carrier frequency shifts of -4, -2, 0, 2, and 4 kHz. Finally, it plots the correlograms. The correlograms for -2 kHz and 2 kHz look identically. Hence, if we do not look at the phase of any of the signals, I would say that we cannot decide whether the original "input signal" has a carrier frequency of -2 kHz or 2 kHz. (The code roughly follows [this implementation](https://github.com/perrysou/GNSS_SDR/blob/master/acquisition.m)).
```python
import numpy as np
import matplotlib.pyplot as plt


# Signal parameters
intermediate_frequency = 0.0
sampling_frequency = 4.092e6
doppler_shift = 2e3

# Time vector (1 ms)
time = np.arange(4092) / sampling_frequency

# # For simplicity, assume that code is just always +1
# code = np.ones(4092)
# Generate random binary code with frequency 1.023 MHz and duration 1 ms
code = np.repeat(np.random.random_integers(0, 1, 1023), 4) * 2 - 1

# Generate fake real input signal with +2 kHz shift
carrier = np.sin(2.0 * np.pi * (intermediate_frequency + doppler_shift) * time)
signal = carrier * code
# Apply random phase shift
signal = np.roll(signal, np.random.random_integers(0, 4092))

# Generate signals with -4, -2, 0, 2, 4 kHz shifts to compare with
carrier_replicas = [
    np.exp(1j * 2.0 * np.pi * (intermediate_frequency + shift) * time)
    for shift in [-4e3, -2e3, 0, 2e3, 4e3]
    ]

# Multiply input signal with carrier candidates ("wipe-off carrier")
multiplied_signals = [signal * carrier_replica
                      for carrier_replica in carrier_replicas]

# Correlate with code in frequency domain
correlations = [np.abs(
    np.fft.ifft(np.conjugate(np.fft.fft(code)) * np.fft.fft(multiplied_signal))
    )**2 for multiplied_signal in multiplied_signals]

# Plot correlations
plt.figure()
for k in range(5):
    plt.subplot(1, 5, k+1)
    plt.plot(correlations[k])
    plt.ylim([0, 5e6])
    plt.title(f"frequency shift: {[-4e3, -2e3, 0, 2e3, 4e3][k]} Hz")
    plt.xlabel("code shift")
plt.show()
```
![correlations](https://user-images.githubusercontent.com/57848364/152770704-b5659282-8889-4077-8786-67e85b4b49c2.png)

Finally, note that the intermediate frequency in the demo is not 0 Hz and, therefore, you do not face this issue.

**Reply by ZiglioUK** [reply]
Hi @JonasBchrt 

the reason why you see two doppler solutions is because you've applied a real sine to modulate the signal, therefore you get two replicas, one for positive, one for negative doppler.
If you use a complex sinusoid (like you do for demodulation), you only get one:

```
# Generate fake real input signal with +2 kHz shift
#carrier = np.sin(2.0 * np.pi * (intermediate_frequency + doppler_shift) * time)
carrier = np.exp(1j * 2.0 * np.pi * (intermediate_frequency + doppler_shift) * time)
```

![image](https://user-images.githubusercontent.com/1219039/152871062-56d4b750-a710-46c8-95f6-508d566169bd.png)

**Reply by JonasBchrt** [reply]
...but this does not work for a real input signal, right?

Maybe I should have written sth like
```python
# Generate fake real input signal with +2 kHz shift
carrier = np.exp(1j * (2.0 * np.pi * (intermediate_frequency + doppler_shift) * time + 2.0 * np.pi * np.random.rand()))
# Use only real carrier
carrier = np.real(carrier)
```
to make clear that the real input signal is the real part of a complex signal. (The last term in the exponent is just there to introduce a random phase shift.)

**Reply by ZiglioUK** [reply]
sorry, why are you interested in only the real part of a complex signal? a complex signal that is downconverted from IF to baseband is still complex

**Reply by JonasBchrt** [reply]
Well, the point is probably not what I am interested in, but what the student who originally asked the question is interested in. I do not know at which dataset he is looking, but he had some trouble getting the cold snapshot algorithm to work. One of several guesses that came to my mind was that he might be looking at a dataset with just real signals and a centre frequency of 0 Hz. I know that datasets like this can be found online.

SnapperGPS also just uses real signals, but that is a completely different story. We want to reduce the amount of collected data there and are not really interested in estimating the Doppler.

**Reply by ZiglioUK** [reply]
I think @willhakkrd asked the question because [above](https://github.com/JonasBchrt/snapshot-gnss-algorithms/discussions/2#discussioncomment-2049017) you had mentioned `"A standard acquisition routine cannot distinguish between a shift of -200 Hz and +200 Hz"`. Also in your paper you wrote `"if the IF is exactly zero, then a carrier frequency that differs by -50 Hz from the IF is indistinguishable from one that deviates by +50 Hz"`.
Now when I read this I was also puzzled. I can't find in the paper you mentioning of dealing with real signals rather than complex.
In fact, when I look at the [acquisition](https://github.com/JonasBchrt/snapshot-gnss-algorithms/blob/422d679237b6aa0b891943a0b1bab59fc85e5dfe/eph_util.py#L2517) algorithm all I see is processing of a complex signal, I'm confused about where exactly you're dealing with a real signal.

**Reply by JonasBchrt** [reply]
Yeah, it was not very thoughtful to write "A standard acquisition" here. Thanks for your comments.

**Reply by JonasBchrt** [reply]
Maybe my paper would have benefited from you being on the board of reviewers ;)

**Reply by ZiglioUK** [reply]
I'm just happy about having received a mention, thanks for that! I would have had a "sentimental" conflict of interest there... But I'm following the development with curiosity!

**Reply by JonasBchrt** [reply]
Anyways, SnapperGPS deals with binary real signals, 1 bit per sample. I think it is at least mentioned in Section 2, but probably could be more prominent.

**Reply by JonasBchrt** [reply]
My code in the comments above as well as the two acquisition functions in this repo take real inputs as well as complex inputs.
If you feed a complex signal into the code stub above, then the the correlograms look different for -2 kHz and 2 kHz. If you feed an unmodified real signal into it, then they look the same.
